kythe/docs/rfc/2909.md

```diff
@@ -0,0 +1,145 @@
+
+# Proposal
+
+Split serving table ProtocolBuffer values into their component fields, making each an independent key-value entry.  Single lookup API calls become scans over a key prefix; paging becomes trivial.  Due the characteristics of LevelDB/SSTables, the split key-value entries will most likely be read together (see SSTable block encoding), keys will be prefix-encoded, and there should be no significant performance hit for reads.  On the contrary, since the Kythe API gives users the ability to request partial data, the split key-value entries allows the server to possibly skip over large segments of data while scanning and save the cost of decoding for a performance improvement.  Examples include a user requesting only small span of a file's decorations or a user requesting only a single kind of cross-references.
+
+More importantly, keeping serving data separate lets the Kythe PostProcessor do fewer large joins.  There will no longer need to be a join of all file decorations to create the single `FileDecorations` proto (likewise for CrossReferences).  Fewer large joins avoids stragglers in the PostProcessor and removes the need for many "limiters" (i.e. cutoff values for the size of outputs).
+
+The scope of this proposal only includes FileDecorations and CrossReferences.  Other services could be made to use similar formats, but the most improvement will be gained from these two APIs.
+
+# Implementation
+
+Essentially each field of an existing ProtocolBuffer value in the serving data becomes zero or more key-value entries depending on whether it's unset, set, or a repeated field.  The bulk of the field data will be represented in the key and the keys will be encoded as [orderedcodes](https://godoc.org/github.com/google/orderedcode).  It's important that each key be unique to fit storage models like LevelDB which are not multimaps (unlike the underlying SSTables).  Tags within keys will indicate how to decode the values themselves and help order the key-value entries.  The ordering of the keys will be chosen as to best fit a performant server, allowing for easy, single-pass scans to retrieve all possible data.
```

Comment:
The constraint that this design works in a monomap data store is important enough that I'd include it in the general section.

Reply:
I added a mention above.